pkg/serviceaccount/jwt_external.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serviceaccount
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+	jose "gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/klog"
+	externalsigner "k8s.io/kubernetes/pkg/serviceaccount/externalsigner/v1beta1"
+)
+
+func ExternalJWTTokenGenerator(iss string, socketPath string) (TokenGenerator, error) {
+	conn, err := grpc.Dial(fmt.Sprintf("unix://%s", socketPath), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	return &externalJWTTokenGenerator{
+		iss:    iss,
+		client: externalsigner.NewJWTKeyServiceClient(conn),
+	}, nil
+}
+
+type externalJWTTokenGenerator struct {
+	iss    string
+	client externalsigner.JWTKeyServiceClient
+}
+
+func (g *externalJWTTokenGenerator) GenerateToken(claims *jwt.Claims, privateClaims interface{}) (string, error) {
+
+	signer := NewRemoteOpaqueSigner(g.client)
+
+	generator, err := JWTTokenGenerator(g.iss, signer.Public().KeyID, signer)
+	if err != nil {
+		return "", err
+	}
+	return generator.GenerateToken(claims, privateClaims)
+}
+
+// NewRemoteOpaqueSigner returns an jose.OpaqueSigner that communicates over the client
+func NewRemoteOpaqueSigner(client externalsigner.JWTKeyServiceClient) jose.OpaqueSigner {
```

Comment:
Make `opequeSigner` exported and return `*OpaqueSigner`.

Reply:
Done